File: android/app/src/main/kotlin/com/example/loc_360/TrackingState.kt
package com.spacewire.circle360

import android.content.Context
import android.location.Location
import java.util.UUID

/**
 * SharedPreferences-backed state shared by the activity, the service and the boot receiver.
 *
 * The service keeps running after the Flutter engine is gone, so this is the only place where
 * "is tracking on" and the upload counters survive. The UI reads it back on resume instead of
 * assuming a fresh start.
 */
object TrackingState {

    private const val PREFS = "loc360_state"

    private const val KEY_TRACKING = "is_tracking"
    private const val KEY_DEVICE_ID = "device_id"
    private const val KEY_ENDPOINT = "upload_endpoint"
    private const val KEY_API_KEY = "upload_api_key"
    private const val KEY_TOKEN = "upload_token"
    private const val KEY_LAST_LAT = "last_lat"
    private const val KEY_LAST_LNG = "last_lng"
    private const val KEY_LAST_ACCURACY = "last_accuracy"
    private const val KEY_LAST_FIX_AT = "last_fix_at"
    private const val KEY_LAST_SUCCESS_AT = "last_success_at"
    private const val KEY_SUCCESS_COUNT = "success_count"
    private const val KEY_FAILURE_COUNT = "failure_count"
    private const val KEY_LAST_ERROR = "last_error"

    private fun prefs(context: Context) =
        context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    fun isTracking(context: Context): Boolean = prefs(context).getBoolean(KEY_TRACKING, false)

    fun setTracking(context: Context, tracking: Boolean) {
        prefs(context).edit().putBoolean(KEY_TRACKING, tracking).apply()
    }

    /** Install-scoped id, reported alongside each fix so multiple devices can be told apart. */
    fun deviceId(context: Context): String {
        val p = prefs(context)
        return p.getString(KEY_DEVICE_ID, null) ?: UUID.randomUUID().toString().also {
            p.edit().putString(KEY_DEVICE_ID, it).apply()
        }
    }

    /**
     * Where uploads go and what authenticates them.
     *
     * The service outlives the Flutter engine, so it cannot ask Dart for the session token when
     * it needs one — Dart pushes it here at sign-in instead, and this is the only copy that
     * survives a force-quit or a reboot.
     */
    data class Upload(val endpoint: String, val apiKey: String, val token: String)

    fun setUpload(context: Context, endpoint: String, apiKey: String, token: String) {
        prefs(context).edit()
            .putString(KEY_ENDPOINT, endpoint)
            .putString(KEY_API_KEY, apiKey)
            .putString(KEY_TOKEN, token)
            .apply()
    }

    fun clearUpload(context: Context) {
        prefs(context).edit()
            .remove(KEY_ENDPOINT)
            .remove(KEY_API_KEY)
            .remove(KEY_TOKEN)
            .apply()
    }

    /** Null until Dart has signed in. A null here means "do not upload", never "upload anonymously". */
    fun upload(context: Context): Upload? {
        val p = prefs(context)
        val endpoint = p.getString(KEY_ENDPOINT, null)
        val token = p.getString(KEY_TOKEN, null)
        if (endpoint.isNullOrBlank() || token.isNullOrBlank()) return null
        return Upload(endpoint, p.getString(KEY_API_KEY, null).orEmpty(), token)
    }

    fun recordFix(context: Context, location: Location) {
        prefs(context).edit()
            .putFloat(KEY_LAST_LAT, location.latitude.toFloat())
            .putFloat(KEY_LAST_LNG, location.longitude.toFloat())
            .putFloat(KEY_LAST_ACCURACY, location.accuracy)
            .putLong(KEY_LAST_FIX_AT, System.currentTimeMillis())
            // Floats lose precision past ~7 digits, so keep the authoritative values as strings.
            .putString("last_lat_exact", location.latitude.toString())
            .putString("last_lng_exact", location.longitude.toString())
            .apply()
    }

    fun recordUploadSuccess(context: Context) {
        val p = prefs(context)
        p.edit()
            .putLong(KEY_LAST_SUCCESS_AT, System.currentTimeMillis())
            .putInt(KEY_SUCCESS_COUNT, p.getInt(KEY_SUCCESS_COUNT, 0) + 1)
            .remove(KEY_LAST_ERROR)
            .apply()
    }

    fun recordUploadFailure(context: Context, error: String) {
        val p = prefs(context)
        p.edit()
            .putInt(KEY_FAILURE_COUNT, p.getInt(KEY_FAILURE_COUNT, 0) + 1)
            .putString(KEY_LAST_ERROR, error.take(200))
            .apply()
    }

    /** Snapshot in the exact shape the Dart layer expects from `getStatus()`. */
    fun snapshot(context: Context): Map<String, Any?> {
        val p = prefs(context)
        val fixAt = p.getLong(KEY_LAST_FIX_AT, 0L)
        return mapOf(
            "isTracking" to p.getBoolean(KEY_TRACKING, false),
            "latitude" to p.getString("last_lat_exact", null)?.toDoubleOrNull(),
            "longitude" to p.getString("last_lng_exact", null)?.toDoubleOrNull(),
            "accuracy" to if (fixAt > 0) p.getFloat(KEY_LAST_ACCURACY, 0f).toDouble() else null,
            "lastFixAt" to if (fixAt > 0) fixAt else null,
            "lastSuccessAt" to p.getLong(KEY_LAST_SUCCESS_AT, 0L).takeIf { it > 0 },
            "successCount" to p.getInt(KEY_SUCCESS_COUNT, 0),
            "failureCount" to p.getInt(KEY_FAILURE_COUNT, 0),
            "lastError" to p.getString(KEY_LAST_ERROR, null),
            // Lets the UI distinguish "tracking is off" from "tracking is on but going nowhere",
            // which otherwise look identical from Dart.
            "uploadConfigured" to (upload(context) != null),
        )
    }
}

File: ios/Runner/TrackingState.swift
import Foundation
import CoreLocation

/// UserDefaults-backed mirror of the Android `TrackingState`.
///
/// iOS can relaunch the app in the background after a force-quit, at which point nothing from the
/// previous process survives except this. `isTracking` is what tells the relaunched process that
/// it should resume rather than sit idle.
enum TrackingState {

    private static let defaults = UserDefaults.standard

    private enum Key {
        static let tracking = "loc360.isTracking"
        static let deviceId = "loc360.deviceId"
        static let lat = "loc360.lastLat"
        static let lng = "loc360.lastLng"
        static let accuracy = "loc360.lastAccuracy"
        static let fixAt = "loc360.lastFixAt"
        static let successAt = "loc360.lastSuccessAt"
        // The last position the backend accepted, which is what the upload gate measures
        // against. Deliberately separate from the last *fix* above, which is written on every
        // delegate callback and so would always be zero metres away.
        static let sentLat = "loc360.sentLat"
        static let sentLng = "loc360.sentLng"
        static let sentAt = "loc360.sentAt"
        static let successCount = "loc360.successCount"
        static let failureCount = "loc360.failureCount"
        static let lastError = "loc360.lastError"
        static let endpoint = "loc360.uploadEndpoint"
        static let apiKey = "loc360.uploadApiKey"
        static let token = "loc360.uploadToken"
    }

    /// Where uploads go and what authenticates them.
    ///
    /// iOS can relaunch this process in the background with no Flutter engine attached, so the
    /// uploader cannot ask Dart for the session token when it needs one. Dart pushes it here at
    /// sign-in, and this is the only copy that survives a force-quit.
    struct Upload {
        let endpoint: String
        let apiKey: String
        let token: String
    }

    static func setUpload(endpoint: String, apiKey: String, token: String) {
        defaults.set(endpoint, forKey: Key.endpoint)
        defaults.set(apiKey, forKey: Key.apiKey)
        defaults.set(token, forKey: Key.token)
    }

    static func clearUpload() {
        defaults.removeObject(forKey: Key.endpoint)
        defaults.removeObject(forKey: Key.apiKey)
        defaults.removeObject(forKey: Key.token)
        // Dropped with the credential: the next sign-in is a different session, and possibly a
        // different person on a shared handset. Leaving it would let the gate suppress their
        // first upload for a whole heartbeat.
        defaults.removeObject(forKey: Key.sentLat)
        defaults.removeObject(forKey: Key.sentLng)
        defaults.removeObject(forKey: Key.sentAt)
    }

    /// Nil until Dart has signed in. Nil means "do not upload", never "upload anonymously".
    static var upload: Upload? {
        guard
            let endpoint = defaults.string(forKey: Key.endpoint), !endpoint.isEmpty,
            let token = defaults.string(forKey: Key.token), !token.isEmpty
        else { return nil }
        return Upload(
            endpoint: endpoint,
            apiKey: defaults.string(forKey: Key.apiKey) ?? "",
            token: token
        )
    }

    static var isTracking: Bool {
        get { defaults.bool(forKey: Key.tracking) }
        set { defaults.set(newValue, forKey: Key.tracking) }
    }

    /// Install-scoped id so the dummy backend can tell devices apart.
    static var deviceId: String {
        if let existing = defaults.string(forKey: Key.deviceId) { return existing }
        let generated = UUID().uuidString
        defaults.set(generated, forKey: Key.deviceId)
        return generated
    }

    static func recordFix(_ location: CLLocation) {
        defaults.set(location.coordinate.latitude, forKey: Key.lat)
        defaults.set(location.coordinate.longitude, forKey: Key.lng)
        defaults.set(location.horizontalAccuracy, forKey: Key.accuracy)
        defaults.set(Date().timeIntervalSince1970 * 1000, forKey: Key.fixAt)
    }

    /// What the backend last accepted. Nil until the first successful upload.
    static var lastSent: (location: CLLocation, at: Date)? {
        guard defaults.object(forKey: Key.sentAt) != nil else { return nil }
        return (
            CLLocation(
                latitude: defaults.double(forKey: Key.sentLat),
                longitude: defaults.double(forKey: Key.sentLng)
            ),
            Date(timeIntervalSince1970: defaults.double(forKey: Key.sentAt))
        )
    }

    /// Records a position the backend accepted.
    ///
    /// Called only on success, so a failed POST leaves the gate open and the next tick retries
    /// rather than being suppressed until the heartbeat comes round.
    static func recordSent(_ location: CLLocation) {
        defaults.set(location.coordinate.latitude, forKey: Key.sentLat)
        defaults.set(location.coordinate.longitude, forKey: Key.sentLng)
        defaults.set(Date().timeIntervalSince1970, forKey: Key.sentAt)
    }

    static func recordUploadSuccess() {
        defaults.set(Date().timeIntervalSince1970 * 1000, forKey: Key.successAt)
        defaults.set(defaults.integer(forKey: Key.successCount) + 1, forKey: Key.successCount)
        defaults.removeObject(forKey: Key.lastError)
    }

    static func recordUploadFailure(_ error: String) {
        defaults.set(defaults.integer(forKey: Key.failureCount) + 1, forKey: Key.failureCount)
        defaults.set(String(error.prefix(200)), forKey: Key.lastError)
    }

    /// Snapshot in the exact shape the Dart layer expects from `getStatus()`.
    static func snapshot() -> [String: Any?] {
        let fixAt = defaults.double(forKey: Key.fixAt)
        let successAt = defaults.double(forKey: Key.successAt)
        return [
            "isTracking": isTracking,
            "latitude": fixAt > 0 ? defaults.double(forKey: Key.lat) : nil,
            "longitude": fixAt > 0 ? defaults.double(forKey: Key.lng) : nil,
            "accuracy": fixAt > 0 ? defaults.double(forKey: Key.accuracy) : nil,
            "lastFixAt": fixAt > 0 ? Int(fixAt) : nil,
            "lastSuccessAt": successAt > 0 ? Int(successAt) : nil,
            "successCount": defaults.integer(forKey: Key.successCount),
            "failureCount": defaults.integer(forKey: Key.failureCount),
            "lastError": defaults.string(forKey: Key.lastError),
            // No equivalent of Android's battery-optimization whitelist on iOS.
            "batteryOptimized": false,
            // Lets the UI distinguish "tracking is off" from "tracking is on but going nowhere".
            "uploadConfigured": upload != nil,
        ]
    }
}
